### ISSUE TITLE
Bug: [no-unnecessary-condition] incorrectly flags on uninitialized variables

### ISSUE THREAD

**Issue by nstepien** [typescript-eslint/typescript-eslint#6640]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&jsx=true&sourceType=module&code=DYUwLgBA9gDmCWUB2BnAXBArkg1kqA7kgNoC6A3ALABQNAZtgMYLIQDm4A8nIqgBQBKCAG8aEaD2QoIAfhkBeCGSrVxAJ3CY1SCS1QqAvkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRgAjODABfEGqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoENNkJemy0eFYjArugDTi114AhpgDm6VpUjoSRdAA9WfAL4glQA

### Repro Code

```typescript
let options: unknown[];

function getOptions() {
  options ??= [];
  return options;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": 1
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true,
    "target": "esnext"
  }
}
```

### Expected Result

Valid code should not be flagged

### Actual Result

Valid code is flagged

### Additional Info

This started happening with v5.55.0.

**Comment by bradzacher** [comment]
This is similar to #6635 and #6635
`options ?? []` will report because according to the types `options` is never nullish.
Because `options ?? []` reports, `options ??= []` also reports, given that it's syntactic sugar for `options ?? (options = [])`

**Comment by nstepien** [comment]
Maybe it should be fixed in TS itself? 🤔 
TS does flag access to the variable before it's initialized, but does not give the correct type.
I believe it should be able to infer the correct type at such location.

![image](https://user-images.githubusercontent.com/567105/225327046-02a66fe2-8be2-4f6d-a42b-fd74efacbc61.png)
![image](https://user-images.githubusercontent.com/567105/225327118-1e8633c9-2092-497c-95df-df78feb02eb5.png)

**Comment by bradzacher** [comment]
Potentially. This has always been the behaviour in TS - it detects used-before-initialised via control-flow analysis solely, and does not change the types.

We have had to deal with this before in other rules https://github.com/typescript-eslint/typescript-eslint/pull/478

**Comment by chharvey** [comment]
As an aside, the used-before-initialized problem can be corrected with the [definite assignment assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#definite-assignment-assertions) in TS:
```ts
let options: unknown[];

options; // ! Variable 'options' is used before being assigned.(2454)

let opts!: unknown[];
//      ^ “It’s not assigned here, but it’s definitely assigned somewhere.”

opts; // no error
```

**Comment by bradzacher** [comment]
Filed https://github.com/microsoft/TypeScript/issues/53306
It would be great if this could be fixed in TS itself so it reports the correct types.
If not, then we can just generalise our used-before-assigned logic from #478 and reuse it for this rule as well.

**Comment by Josh-Cena** [comment]
This is duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/4513
